_posts/2020-08-24-python-calling-c++.md

```diff
@@ -0,0 +1,167 @@
+
+##Simple Example
+
+So now that we are done with the theoretical aspects of it, let’s get our hands dirty with some coding and see how it can be used through a short example. 
+
+Before we get to the actual coding part let’s see what the prerequisites are to run a program with a boost library: 
+
+1. [Boost](https://www.boost.org/) (version >= 1.3.2) 
+
+2. [Python](http://www.python.org/) (version >= 2.2) 
+
+3. A C++ compiler for your platform, e.g. [GCC](https://gcc.gnu.org/) or [MinGW](http://www.mingw.org/) 
+
+Suppose we have the following C++ API which we want to expose in Python:
+
+```cpp
+#include <string> 
+
+namespace 
+{   
+    // Avoid cluttering the global namespace. 
+
+    // A couple of simple C++ functions that we want to expose to Python. 
+    std::string greet() { return "hello, world"; } 
+    int square(int number) { return number * number; }
+} 
+```
+
+Here is the C++ code for a python module called getting_started1 which exposes the API. 
+
+```cpp
+#include <boost/python.hpp> 
+using namespace boost::python; 
+ 
+BOOST_PYTHON_MODULE(getting_started1) 
+{ 
+
+    // Add regular functions to the module. 
+    def("greet", greet); 
+    def("square", square); 
+} 
+```  
+That's it! If we build this shared library and put it on our PYTHONPATH we can now access our C++ functions from Python. 
+
+```python
+>>> import getting_started1 
+>>> print getting_started1.greet()
+ 
+hello, world 
+
+>>> number = 11 
+>>> print number, '*', number, '=', getting_started1.square(number) 
+
+11 * 11 = 121 
+``` 
+
+So, as you can see from the example above the only additional library required to run our program is the Boost library apart from the regular C++ and Python compiler and all you need to do is compile and build the C++ library and import it in Python as a regular Python library and boom you’re ready to go.
+
+##Exposing classes written in C++ to Python 
+
+Now that we have explored running a simple Hello World program in Python which is written in C++, let’s see how we can do the same with C++ classes through an example. 
+
+Assume we want to expose the below written C++ class to Python.    
+```cpp
+struct World 
+{ 
+    void set(std::string msg) { mMsg = msg; }
+    std::string greet() { return mMsg; }
+    std::string mMsg;
+}; 
+ 
+using namespace boost::python; 
+ 
+BOOST_PYTHON_MODULE(classes) 
+        { 
+                class_<World>("World") 
+                        .def("greet", &World::greet) 
+                        .def("set", &World::set) 
+                ; 
+        }; 
+```  
+The corresponding Python code to call the above C++ class from Python would be: 
+
+```python
+import classes 
+ 
+t = classes.World() 
+t.set("Python says hi to C++.") 
+print (t.greet())
+``` 
+The output of running the above code would be: 
+
+```python
+/Users/adityagupta/PycharmProjects/check/venv/bin/python /Users/adityagupta/PycharmProjects/check/Classes.py 
```

Comment:
remove you local path, probably just: `python Classes.py`

Reply:
Sure